### ISSUE TITLE
Memory Leak when testing Activities

### ISSUE THREAD

**Issue by plackemacher** [robolectric/robolectric#2068]
We have over 1000+ tests using the Robolectric suite and we kept running into out of memory issues or heap space issues on our CI environment as we added more tests which forced us to keep increasing the max heap space allowed. From running a local heap dump, I was able to find that many activities (and fragments through associated activities) were being strongly referenced in memory through the single `WindowManagerGlobal` object. Are we supposed to close or shutdown our activities in a particular way after we're done with them? The samples don't seem to demonstrate a best practice for this or maybe I'm just missing something.

As a temporary measure of sorts, I wrote this method to run after each test so as to release any unnecessary references:

``` java
    @SuppressLint("NewApi")
    @After
    public void resetWindowManager() {
        Class clazz = ReflectionHelpers.loadClass(getClass().getClassLoader(), "android.view.WindowManagerGlobal");
        Object instance = ReflectionHelpers.callStaticMethod(clazz, "getInstance");

        // We essentially duplicate what's in {@link WindowManagerGlobal#closeAll} with what's below.
        // The closeAll method has a bit of a bug where it's iterating through the "roots" but
        // bases the number of objects to iterate through by the number of "views." This can result in
        // an {@link java.lang.IndexOutOfBoundsException} being thrown.
        Object lock = ReflectionHelpers.getField(instance, "mLock");

        ArrayList<Object> roots = ReflectionHelpers.getField(instance, "mRoots");
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized (lock) {
            for (int i = 0; i < roots.size(); i++) {
                ReflectionHelpers.callInstanceMethod(instance, "removeViewLocked",
                        ReflectionHelpers.ClassParameter.from(int.class, i),
                        ReflectionHelpers.ClassParameter.from(boolean.class, false));
            }
        }

        // Views will still be held by this array. We need to clear it out to ensure
        // everything is released.
        ArraySet<View> dyingViews = ReflectionHelpers.getField(instance, "mDyingViews");
        dyingViews.clear();
    }
```

From taking heap dumps at roughly the same point in test completion, the total memory usage in the heap dropped from 1.1GB to 159MB.

**Comment by kingargyle** [comment]
I took the above and put it in our afterTest method for our custom TestLifeCycle implementation.  This did let me get through about 3/4 of our 2200+ test suite in IntelliJ before I started getting out of memory issues and gc issues.

**Comment by mag** [comment]
I'm having a similar problem on a large (~1200 tests) suite after upgrading to 3.0. Using JProfiler I could se lots of ShadowApplication objects being leaked. I was also able to reproduce this behavior in the Deckard project as well, so I don't think it's anything specific to my application or test suite.

The hack above posted by @plackemacher worked on Deckard...it clearly eliminated the leaked ShadowApplication objects. Unfortunately that code only works if compiling against SDK 23, I believe, which means I can't try it (as is) because our project has not yet been fixed up for 23 :cry: 

We tried other variations of @plackemacher's workaround code against Deckard  (which compiles agains 22) but couldn't get the ShadowApplications to stop leaking. We don't have a good understanding (yet) about what is holding onto references....I see lots of weird & long chains of references through java.lang.ref.Finalizer objects, but I don't know if that's a leak or just an un-run finalizer method. That being said, if we compile against 23 and use his code we do in fact see no extra ShadowApplications via Finalizers.

Anyway I'm hoping to restart this conversation to see if anybody has learned a better way around this problem. Thanks.

**Comment by mag** [comment]
I figured out how to make the above workaround usable if you're compiling against < API 22 - use Collection<View> in the second-to-last line. This allowed our test suite to finish with a max memory use of 1.5GB.

**Comment by jaredsburrows** [comment]
@plackemacher in your `tearDown` methods, are you destroying your fragments and activities?

**Comment by mag** [comment]
One of my coworkers found that destroying activities in tearDown was effective in getting the activities GC'd - so I think this will work. But it's a pain in the butt for users...and it sounds like something that Robolectric should be able to achieve automatically. Anybody looking at this?

**Comment by jaredsburrows** [comment]
@mag Thanks for confirming. But it is only 1 line of code, ` But it's a pain in the butt for users...`.

It would be nice to do this automatically.

**Comment by plackemacher** [comment]
@jaredsburrows I have tried tearing down the activities and it seemed ineffective but I may not have done it correctly. What does your tear down method look like for activities? This is the method I used:

```java
    protected final void stopActivity() {
        mActivityController.saveInstanceState(mBundle).pause().stop().destroy();

        mActivity = null;
    }
```

**Comment by jaredsburrows** [comment]
@plackemacher I always do something like:

**Before:**
```
    private ActivityController<MainActivity> controller;
    private MainActivity sut;
    private ShadowActivity shadowActivity;

    @Before
    public void setUp() throws Exception {

        controller = buildActivity(MainActivity.class);
        sut = controller.create().start().resume().visible().get();
        shadowActivity = shadowOf(sut);
}
```

**After:**
```
    @After
    public void tearDown() throws Exception {

        controller.pause().stop().destroy();
    }
```

Besides tearing down the activities, I am also ensuring there are no crashes happening in `pause`, `stop` and `destroy`.

**Comment by kingargyle** [comment]
Calling, activityUnderTest.finish() seems to be all that is necessary in a tearDown method.  After doing this, I was able to get  my test suite to run to completion within IntelliJ which it wasn't able to do before.

**Comment by jaredsburrows** [comment]
Did you try my teardown? Or are you just calling finish at the end of each
test?

**Comment by kingargyle** [comment]
Just calling finish at the end of each test.

**Comment by SimonHaenle** [comment]
We have exactley the same isues. We have 1700 Tests, 6GB Heapspace, and a 50/50 chance that our tests will run. SOmetimes they run, sometimes we get a GCOverhead. 
Finishing activities does not help.

**Comment by jongerrish** [comment]
Please post profiling data.

**Comment by hzsweers** [comment]
We had to pull in that clearing out of dying views as well, and it was blowing up memory usage on our tests. This seems like a major issue

**Comment by mniami** [comment]
I'm agree with previous comments, we have 1623 tests and they're starting failing after 1230 test which looks like the limit for robolectric. As it was said before even if you use activityController.destroy and null activity in tearDown you have still the same issue with memory leak.

**Comment by jaredsburrows** [comment]
@mniami Can some tests be converted to plain junit tests? or mockito runner tests?
Have you tried killing off the activities in the `tearDown`? See: https://github.com/robolectric/robolectric/issues/2068#issuecomment-167858811.

**Comment by mniami** [comment]
@jaredsburrows whole amount of tests is 3450 so half of them are using robolectric the rest is without it.
I'm trying to solve the problem by getting shadowContextProvider and clear the whole hashMap with contentProviders on the end of package tests. If you can help me with that I will be appreciate. 
About killing activities in tearDown - yes we do that, destroying it by activityController and finishing on the very end.

**Comment by connected-jallen** [comment]
@plackemacher Your teardown was a lifesaver.

**Comment by jiahaoliuliu** [comment]
Based on the comment of @jaredsburrows I have created the follow base test. You can see the example of usage here:
https://github.com/jiahaoliuliu/robolectricForMarshmallow/tree/baseTestRunner

```java
package com.jiahaoliuliu.robolectricformarshmallow;

import android.app.Activity;

import com.jiahaoliuliu.robolectricformarshmallow.shadow.MainControllerShadow;

import org.junit.After;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.util.ActivityController;

/**
 * Created by jiahao on 19/09/16.
 */
@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, manifest = Config.NONE, application = FoolApplication.class
        , shadows = {MainControllerShadow.class}
        )
public class BaseTest <T extends Activity>{

    private Class mClazz;
    private ActivityController<T> mActivityController;

    public BaseTest(Class<T> clazz) {
        this.mClazz = clazz;
    }

    /**
     * Get the activity. This method is not set inside the constructor but outside
     * allow the tests control better when the activities will be set.
     * @return
     *      The activity already visible
     */
    protected T getActivity() {
        this.mActivityController =
                Robolectric.buildActivity(mClazz).create().start().postCreate(null).resume().visible();
        return mActivityController.get();
    }

    @After
    public void tearDown() throws Exception{
        if (mActivityController != null) {
            mActivityController.pause().stop().destroy();
        }
    }
}
```
Usage:

```java
package com.jiahaoliuliu.robolectricformarshmallow;

import com.jiahaoliuliu.robolectricformarshmallow.shadow.SettingsShadow;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.robolectric.annotation.Config;

import static org.junit.Assert.*;

/**
 * Created by Jiahao on 6/30/16.
 */
@Config(shadows = {SettingsShadow.class})
public class MainActivityTest extends BaseTest<MainActivity>{
    private MainActivity mMainActivity;

    public MainActivityTest() {
        super(MainActivity.class);
    }

    @Before
    public void setUp() throws Exception {
        // Getting the activity.
        mMainActivity = getActivity();
    }

    @After
    public void tearDown() throws Exception {
        super.tearDown();
    }

    @Test
    public void testOnCreate() throws Exception {
        // Simple test to know that it works
        assertTrue(true);

        // The debug is always set in the shadows
        String actionBarTitle = mMainActivity.getSupportActionBar().getTitle().toString();
        assertEquals(actionBarTitle, mMainActivity.getResources().getString(R.string.app_name_debugging));
    }
}
```

**Comment by jaredsburrows** [comment]
@jiahaoliuliu Nice! I have something similar:

```
/**
 * Robolectric + JUnit Tests.
 *
 * @author <a href="mailto:jaredsburrows@gmail.com">Jared Burrows</a>
 */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = Build.VERSION_CODES.M)
public abstract class RoboTestBase {
  // Android Related
  protected final Context context = RuntimeEnvironment.application;
  protected final Application application = RuntimeEnvironment.application;
  protected final Resources resources = context.getResources();

  @Before @Override public void setUp() throws Exception {
    super.setUp();

    ShadowApplication.getInstance().declareActionUnbindable("com.google.android.gms.analytics.service.START");
  }

  @After @Override public void tearDown() throws Exception {
    super.tearDown();

    // https://github.com/robolectric/robolectric/issues/1700
    // https://github.com/robolectric/robolectric/issues/2068
    // https://github.com/robolectric/robolectric/issues/2584
    this.resetBackgroundThread();
    this.resetWindowManager();
  }

  public void finishThreads() {
    runBackgroundTasks();
    flushForegroundThreadScheduler();
    flushBackgroundThreadScheduler();
    runUiThreadTasksIncludingDelayedTasks();
  }

  // https://github.com/robolectric/robolectric/pull/1741
  private void resetBackgroundThread() throws Exception {
    final Class<?> btclass = Class.forName("com.android.internal.os.BackgroundThread");
    final Object backgroundThreadSingleton = ReflectionHelpers.getStaticField(btclass, "sInstance");
    if (backgroundThreadSingleton != null) {
      btclass.getMethod("quit").invoke(backgroundThreadSingleton);
      ReflectionHelpers.setStaticField(btclass, "sInstance", null);
      ReflectionHelpers.setStaticField(btclass, "sHandler", null);
    }
  }

  // https://github.com/robolectric/robolectric/issues/2068#issue-109132096
  @SuppressLint("NewApi")
  private void resetWindowManager() {
    final Class<?> clazz = ReflectionHelpers.loadClass(this.getClass().getClassLoader(), "android.view.WindowManagerGlobal");
    final Object instance = ReflectionHelpers.callStaticMethod(clazz, "getInstance");

    // We essentially duplicate what's in {@link WindowManagerGlobal#closeAll} with what's below.
    // The closeAll method has a bit of a bug where it's iterating through the "roots" but
    // bases the number of objects to iterate through by the number of "views." This can result in
    // an {@link java.lang.IndexOutOfBoundsException} being thrown.
    final Object lock = ReflectionHelpers.getField(instance, "mLock");

    final List<Object> roots = ReflectionHelpers.getField(instance, "mRoots");
    //noinspection SynchronizationOnLocalVariableOrMethodParameter
    synchronized (lock) {
      for (int i = 0; i < roots.size(); i++) {
        ReflectionHelpers.callInstanceMethod(instance, "removeViewLocked",
          ReflectionHelpers.ClassParameter.from(int.class, i),
          ReflectionHelpers.ClassParameter.from(boolean.class, false));
      }
    }

    // Views will still be held by this array. We need to clear it out to ensure
    // everything is released.
    final ArraySet<View> dyingViews = ReflectionHelpers.getField(instance, "mDyingViews");
    dyingViews.clear();
  }
}
```

**Comment by swayangjit** [comment]
```javascript
 android.testOptions.unitTests.all {
    maxParallelForks = 1
    forkEvery = 1
}
```
Add this in build.gradle(app level).It will not work if it is added in outside app level build.gradle .

**Comment by jaredsburrows** [comment]
@swayangjit What does this have to do with memory leaks?

**Comment by jongerrish** [comment]
New JVM every invocation so memory leaks are not an issue?

**Comment by jaredsburrows** [comment]
@swayangjit @jongerrish `maxParallelForks` defaults to 1 and `forkEvery` defaults to 0. I just do not add those.

See here: https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html#org.gradle.api.tasks.testing.Test:maxParallelForks
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html#org.gradle.api.tasks.testing.Test:forkEvery

**Comment by swayangjit** [comment]
@jaredsburrows  @jongerrish It will not avoid memory leak but it will  ensure all  your testcase will execute successfully

**Comment by gkylafas** [comment]
@xian unfortunately #2838 does not seem to fix the memory leaks. Up until now I used @jaredsburrows' solution, as described in https://github.com/robolectric/robolectric/issues/2068#issuecomment-248830370.
If I upgrade to robolectric-3.2.2, which contains the fix #2838 in, and disable `resetWindowManager()` from the above comment, then my tests start failing again with out of memory errors.
